[RNMobile] Allow dragging from block toolbar

fluiddot opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Currently, the dragging gesture for a selected block is only allowed by long-pressing over the content. This PR adds the option to also initiate the dragging by long-pressing over the empty space within the block toolbar.

<img src=https://user-images.githubusercontent.com/14905380/167117067-b0ff44b2-3fdb-4ca3-8c23-89b8ef52ba82.png width=300>

Additionally, it also addresses an issue related to nested blocks where up/down arrow buttons within the block toolbar can't be used.

https://user-images.githubusercontent.com/14905380/167118366-7f0d99c5-6b11-41f5-90b3-aed814336ce2.mp4

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Addresses https://github.com/wordpress-mobile/gutenberg-mobile/issues/4791.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- The component `BlockDraggable` is being rendered in `BlockMobileToolbar` wrapping the spacer view, and the configuration values for the dragging are passed via props from the parent component `BlockListBlock`.
- For addressing the issue on nested blocks, I updated the logic related to calculating the `enable` prop. The idea is that for blocks that contain nested blocks, only the content (i.e. the nested blocks) triggers the dragging.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Root block
1. Open a post/page.
2. Add an Image block with an image.
3. Select the Image block.
4. Long-press over the up/down buttons located in the block toolbar.
5. Observe that a picker is displayed with expected options.
6. Long-press over the content of the Image block (i.e. over the image).
7. Observe that the drag mode is enabled.
8. Long-press over the empty space in the block toolbar.
9. Observe that the drag mode is enabled. 

### Block with nested blocks
1. Open a post/page.
2. Add a Group block.
3. Select the Group block.
4. Add different blocks into the Group block.
5. Check that the currently selected block is the Group block (i.e. not a nested block).
6. Long-press over the up/down buttons located in the block toolbar.
7. Observe that a picker is displayed with expected options.
10. Long-press over the content of the Group block.
11. Observe that the drag mode is enabled and the block being dragged is the Group block.
12. Long-press over the empty space in the block toolbar.
13. Observe that the drag mode is enabled and the block being dragged is the Group block.

### Nested block
1. Open a post/page.
2. Add a Group block.
3. Select the Group block.
4. Add different blocks into the Group block.
5. Select a nested block.
6. Long-press over the up/down buttons located in the block toolbar.
7. Observe that a picker is displayed with expected options.
8. Long-press over the content of the nested block.
9. Observe that the drag mode is enabled and the block being dragged is the Group block.
10. Long-press over the empty space in the block toolbar.
11. Observe that the drag mode is enabled and the block being dragged is the Group block.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/14905380/167120692-5ba7b406-2ecc-4c81-8460-a73dbf1fac62.mp4

fluiddot committed:
Enable dragging from block toolbar

fluiddot committed:
Ensure root block is dragged for nested blocks

geriux commented on packages/block-editor/src/components/block-mobile-toolbar/index.native.js:
```diff
@@ -73,7 +76,12 @@ const BlockMobileToolbar = ( {
+				enabled={ draggingEnabled }
+				clientId={ draggingClientId }
+			>
+				{ () => <View style={ styles.spacer } /> }
```
Nice that we already had this "spacer" to attach the `BlockDraggable` to 😄

geriux commented:
This is looking great! Nice job finding a nice approach to support dragging from the toolbar and also for finding that bug with the nested blocks and the up/down arrow buttons. Related to the last one I did find an issue nesting the following blocks:

**Group block > Columns > Column > Image**

When trying to long-press on the up/down arrows in the Column block the dragging mode gets activated:

https://user-images.githubusercontent.com/4885740/167138581-36b40830-5225-4f8b-a0d9-6cdc6206ef06.mov

fluiddot committed:
Add draggingClientId prop to BlockDraggable

With this change the `BlockDraggable` component might receive two different client ids, one is the client id of the block where the component is rendered, and the other (which is optional) is used for identifying the block to be dragged.

fluiddot committed:
Set dragging always enabled for block toolbar

The block toolbar is only visible when the block is selected so it's safe to allow dragging in all cases, as it won't affect other UI elements.

fluiddot committed:
Update dragging enabled calculation

In order to prevent issues related to disabling the long-press gesture in elements within the block UI, the logic for enabling the dragging has been updated. Now it will enabled in the following cases:
- The block doesn't have inner blocks. This applies to root blocks and nested blocks without further nested blocks.
- Blocks with nested blocks if the block is selected.
- Blocks with nested blocks if none of the nested blocks is selected.

fluiddot commented:
> This is looking great! Nice job finding a nice approach to support dragging from the toolbar and also for finding that bug with the nested blocks and the up/down arrow buttons. Related to the last one I did find an issue nesting the following blocks:
> 
> **Group block > Columns > Column > Image**
> 
> When trying to long-press on the up/down arrows in the Column block the dragging mode gets activated:
> 
>  Nested.mov

@geriux good catch! I've just pushed some changes to address this case. It implied a small refactor, but hopefully, it will cover all cases. Let me know if you could take a look, thanks 🙇 .

### Two client id props in `BlockDraggable`

Related commit: https://github.com/WordPress/gutenberg/pull/40886/commits/998b81e84c685920ae37a69d3f7ad62766b9427e

The idea is that we'll keep passing the client id of the block via the `clientId` prop, and include a new one (`draggingClientId`) that will be used for determining which block should be dragged. This is particularly useful for nested blocks, as they will make the root parent block be dragged instead.

### Dragging is always enabled in the block toolbar

Related commit: https://github.com/WordPress/gutenberg/pull/40886/commits/adbc1d69daf3871ccc899b98a88c70e982e13375

I realized that it's not necessary to enable/disable the `BlockDraggable` component when is used within the block toolbar, as it's only visible when the block is selected. Hence, it won't impact other UI elements within the block.

### Update enable dragging calculation

Related commit: https://github.com/WordPress/gutenberg/pull/40886/commits/c3e65767a59ebb4ea7215159bf48efb134c83f78

In relation to the issue you spotted, I updated this logic to cover all cases when dragging nested blocks and blocks with nested blocks. In order to test these changes, I share here the different cases that should be checked:
1. Block with nested blocks and **not selected**:
    - Long-press over the content should enable drag mode.
2. Block with nested blocks and **selected**:
    - Long-press over the content should enable drag mode.
    - Long-press over the block toolbar should enable drag mode.
    - Long-press over the up/down arrow buttons should display the picker.
    - Long-press over a **not selected block** should enable drag mode.
3. Nested block (located at **first level**):
    - Long-press over the content should enable drag mode.
    - Long-press over the block toolbar should enable drag mode.
    - Long-press over the up/down arrow buttons should display the picker.
    - Long-press over a **not selected block outside the parent block** should enable drag mode.
    - Long-press over a **not selected block within the same parent block** should enable drag mode.
4. Nested block (located at a **second or deeper level**):
    - Long-press over the content should enable drag mode.
    - Long-press over the block toolbar should enable drag mode.
    - Long-press over the up/down arrow buttons should display the picker.
    - Long-press over a **not selected block outside the parent block** should enable drag mode.
    - Long-press over a **not selected block within the same parent block** should enable drag mode.

**Example:**

https://user-images.githubusercontent.com/14905380/167168018-ff317dac-2f48-4aa4-8f37-9ce4566a4ba8.mp4

geriux reviewed (approved):
LGTM! Thanks for making the changes, this now works as expected 🚀 

I tested it on both iOS and Android.

Nice work!
